File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-0.js
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    largestRectangleArea(heights) {
        let max = 0;
        const stack = [];
        for(let i=0; i<heights.length; i++) {
            const h = heights[i];
            let start = i;
            while(stack.length && stack.at(-1).height > h) {
                const {start:s,height} = stack.pop();
                max = Math.max(max, height * (i-s));
                start = s
            }
            stack.push({start,height:h});
        }
        for(const {start,height} of stack) {
            max = Math.max(max,height * ( heights.length - start));
        }
        return max
    }
}

File: Data Structures & Algorithms/max-water-container/submission-0.go
func maxArea(heights []int) int {
    l,r := 0, len(heights)-1
    area := 0
    for l < r {
        lh, rh := heights[l], heights[r]
        a := (r-l) * min(lh,rh)
        area = max(a,area)
        if lh == rh {
            if heights[l+1] > heights[r-1] {
                l++
            } else {
                r--
            }
        } else if lh < rh {
            l++
        } else if rh < lh {
            r--
        }

    }
    return area
}
